apis/python/src/tiledb/vector_search/object_readers/directory_reader.py

```diff
@@ -8,6 +8,223 @@
+                        file_uri,
+                        include=include,
+                        exclude=exclude,
+                        suffixes=suffixes,
+                        max_files=max_files,
+                    )
+
+                else:
+                    match = match_uri(
+                        search_uri=search_uri,
+                        file_uri=file_uri,
+                        include=include,
+                        exclude=exclude,
+                        suffixes=suffixes,
+                    )
+                    if match:
+                        results.append(file_uri)
+
+        # Add one trailing slash to search_uri
+        search_uri = search_uri.rstrip("/") + "/"
+        results = []
+        find(
+            search_uri,
+            include=include,
+            exclude=exclude,
+            suffixes=suffixes,
+            max_files=max_files,
+        )
+        if max_files is not None:
+            return results[:max_files]
+        else:
+            return results
+
+
+def find_uris_tiledb_group(
+    search_uri: str,
+    config: Optional[Mapping[str, Any]] = None,
+    include: str = "*",
+    exclude: Sequence[str] = ["[.]*", "*/[.]*"],
+    suffixes: Optional[Sequence[str]] = None,
+    max_files: Optional[int] = None,
+    verbose: bool = False,
+) -> Sequence[str]:
+    with tiledb.scope_ctx(config):
+
+        def find(
+            uri: str,
+            path: str,
+            include: str = "*",
+            exclude: Sequence[str] = (),
+            suffixes: Optional[Sequence[str]] = None,
+            max_files: Optional[int] = None,
+        ):
+            # Stop if we have found max_count files
+            if max_files is not None and len(results) >= max_files:
+                return
+            obj_type = tiledb.object_type(uri)
+            if obj_type == "array":
+                match = match_uri(
+                    search_uri="/",
+                    file_uri=path,
+                    include=include,
+                    exclude=exclude,
+                    suffixes=suffixes,
+                )
+                if match:
+                    results.append(uri)
+            elif obj_type == "group":
+                grp = tiledb.Group(uri)
+                for child in list(grp):
+                    find(
+                        uri=child.uri,
+                        path=f"{path}/{child.name}",
+                        include=include,
+                        exclude=exclude,
+                        suffixes=suffixes,
+                        max_files=max_files,
+                    )
+
+        results = []
+        find(
+            uri=search_uri,
+            path="",
+            include=include,
+            exclude=exclude,
+            suffixes=suffixes,
+            max_files=max_files,
+        )
+        if max_files is not None:
+            return results[:max_files]
+        else:
+            return results
+
+
+def find_uris_aws(
```

Comment:
We should avoid this and use VFS and vfs recursive listing so we are cross platform compatible.

Reply:
This is added for performance reasons. Listing large S3 directories and checking individual files using VFS takes a lot of time. This is also similar to the approach used in [VCF](https://github.com/TileDB-Inc/TileDB-Cloud-Py/blob/main/src/tiledb/cloud/vcf/ingestion.py#L438) for listing S3. I avoided using the `aws` command line util here and switched to `boto3`.

We could add an option to use `find_uris_vfs` for S3 if you think there is going to be a compatibility issue.

Reply:
If there is performance, open a item and lets fix things. You can also use the ls recursive support. VCF does not use aws cli or boto3 by default and was migrated away from it with plans to remove it. Its unused code by default.

As the `find_uris_aws` is written today it also has issues with chaining roles and using a aws file config. The core library supports all of this, lets avoid re-implementing specific functionality already covered in a cross platform way by the core library.

Reply:
Used `find_uris_vfs` for S3. 

What is the vfs ls recursive support? I only see non-recursive ls if TileDB-Py https://github.com/TileDB-Inc/TileDB-Py/blob/5cfefdd0d72a90fa4e16fb2f1107cb09a86b2958/tiledb/vfs.py#L290